driver-core/src/main/com/mongodb/internal/operation/ClientBulkWriteOperation.java

```diff
@@ -258,6 +329,80 @@ private Integer executeBatch(
+                            retryState.attach(AttachmentKeys.maxWireVersion(), connectionDescription.getMaxWireVersion(),
+                                            true)
+                                    .attach(AttachmentKeys.commandDescriptionSupplier(), () -> BULK_WRITE_COMMAND_NAME,
+                                            false);
+                            ClientBulkWriteCommand bulkWriteCommand = createBulkWriteCommand(
+                                    retryState, effectiveRetryWrites, effectiveWriteConcern, sessionContext,
+                                    unexecutedModels, batchEncoder,
+                                    () -> retryState.attach(AttachmentKeys.retryableCommandFlag(), true, true));
+                            executeBulkWriteCommandAndExhaustOkResponseAsync(
+                                    retryState, connectionSource, connection, bulkWriteCommand, effectiveWriteConcern,
+                                    operationContext,
+                                    resultCallback);
+                        }));
+
+        beginAsync()
+                .<ExhaustiveClientBulkWriteCommandOkResponse>thenSupply(callback -> {
+                    retryingBatchExecutor.get(callback);
+                })
+                .<Integer>thenApply((bulkWriteCommandOkResponse, callback) -> {
+                    callback.complete(resultAccumulator.onBulkWriteCommandOkResponseOrNoResponse(
+                            batchStartModelIndex, bulkWriteCommandOkResponse, batchEncoder.intoEncodedBatchInfo()));
+                }).onErrorIf(throwable -> true, (t, callback) -> {
+                    if (t instanceof MongoWriteConcernWithResponseException) {
+                        callback.complete(resultAccumulator.onBulkWriteCommandOkResponseWithWriteConcernError(
+                                batchStartModelIndex, (MongoWriteConcernWithResponseException) t, batchEncoder.intoEncodedBatchInfo()));
+                    } else if (t instanceof MongoCommandException) {
+                        resultAccumulator.onBulkWriteCommandErrorResponse((MongoCommandException) t);
+                        callback.completeExceptionally(t);
+                    } else if (t instanceof MongoException) {
+                        // The server does not have a chance to add "RetryableWriteError" label to `e`,
+                        // and if it is the last attempt failure, `RetryingSyncSupplier` also may not have a chance
+                        // to add the label. So we do that explicitly.
+                        shouldAttemptToRetryWriteAndAddRetryableLabel(retryState, t);
+                        resultAccumulator.onBulkWriteCommandErrorWithoutResponse((MongoException) t);
+                        callback.completeExceptionally(t);
+                    } else {
+                        callback.completeExceptionally(t);
+                    }
+                }).finish(finalCallback);
```

Comment:
Let's format according to the formatting recommended by the `AsyncRunnable` docs, because that way the code looks more similar to the sync counterpart:

```suggestion
        beginAsync().<ExhaustiveClientBulkWriteCommandOkResponse>thenSupply(callback -> {
            retryingBatchExecutor.get(callback);
        }).<Integer>thenApply((bulkWriteCommandOkResponse, callback) -> {
            callback.complete(resultAccumulator.onBulkWriteCommandOkResponseOrNoResponse(
                    batchStartModelIndex, bulkWriteCommandOkResponse, batchEncoder.intoEncodedBatchInfo()));
        }).onErrorIf(throwable -> true, (t, callback) -> {
            if (t instanceof MongoWriteConcernWithResponseException) {
                callback.complete(resultAccumulator.onBulkWriteCommandOkResponseWithWriteConcernError(
                        batchStartModelIndex, (MongoWriteConcernWithResponseException) t, batchEncoder.intoEncodedBatchInfo()));
            } else if (t instanceof MongoCommandException) {
                resultAccumulator.onBulkWriteCommandErrorResponse((MongoCommandException) t);
                callback.completeExceptionally(t);
            } else if (t instanceof MongoException) {
                // The server does not have a chance to add "RetryableWriteError" label to `e`,
                // and if it is the last attempt failure, `RetryingSyncSupplier` also may not have a chance
                // to add the label. So we do that explicitly.
                shouldAttemptToRetryWriteAndAddRetryableLabel(retryState, t);
                resultAccumulator.onBulkWriteCommandErrorWithoutResponse((MongoException) t);
                callback.completeExceptionally(t);
            } else {
                callback.completeExceptionally(t);
            }
        }).finish(finalCallback);
```